pkg/compiler/namespace.go

```diff
@@ -0,0 +1,97 @@
+package compiler
+
+import (
+	"github.com/MontFerret/ferret/pkg/runtime/core"
+	"github.com/pkg/errors"
+	"regexp"
+	"strings"
+)
+
+var fnNameValidation = regexp.MustCompile("^[a-zA-Z]+[a-zA-Z0-9_]*(::[a-zA-Z]+[a-zA-Z0-9_]*)*$")
+
+const emptyNS = ""
+const separator = "::"
+
+type NamespaceBuilder struct {
+	funcs core.Functions
+	name  string
+}
+
+func newRootNamespace() *NamespaceBuilder {
+	ns := new(NamespaceBuilder)
+	ns.funcs = make(core.Functions)
+
+	return ns
+}
+
+func newNamespace(funcs core.Functions, name string) *NamespaceBuilder {
+	return &NamespaceBuilder{funcs, strings.ToUpper(name)}
+}
+
+func (rs *NamespaceBuilder) Namespace(name string) core.Namespace {
+	return newNamespace(rs.funcs, rs.makeFullName(name))
+}
+
+func (rs *NamespaceBuilder) RegisterFunction(name string, fun core.Function) error {
+	nsName := rs.makeFullName(name)
+	_, exists := rs.funcs[nsName]
+
+	if exists {
+		return errors.Errorf("function already exists: %s", name)
+	}
+
+	// validation the name
+	if strings.Contains(name, separator) {
```

Comment:
`fnNameValidation.MatchString(nsName)` enough for validation because `fnNameValidation` match such cases as `FUNC`, `NAMESPACE::FUNC` etc. and doesn't match `FUNC::`

Reply:
So I think it can be removed